File: server/models/pet.py
import uuid
from datetime import datetime, date, timezone

from sqlalchemy import String, DateTime, Date, ForeignKey, Enum as SAEnum, Float
from sqlalchemy.orm import Mapped, mapped_column, relationship
import enum

from server.database.base import Base


def _uuid() -> str:
    return str(uuid.uuid4())


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class PetSpecies(str, enum.Enum):
    DOG = "dog"
    CAT = "cat"
    BIRD = "bird"
    RABBIT = "rabbit"
    FISH = "fish"
    REPTILE = "reptile"
    RODENT = "rodent"
    OTHER = "other"


class PetGender(str, enum.Enum):
    MALE = "male"
    FEMALE = "female"
    UNKNOWN = "unknown"


class Pet(Base):
    __tablename__ = "pets"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=_uuid)
    owner_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )

    name: Mapped[str] = mapped_column(String(80), nullable=False)
    species: Mapped[PetSpecies] = mapped_column(
        SAEnum(PetSpecies, name="pet_species"), nullable=False
    )
    breed: Mapped[str | None] = mapped_column(String(120), nullable=True)
    gender: Mapped[PetGender] = mapped_column(
        SAEnum(PetGender, name="pet_gender"), default=PetGender.UNKNOWN, nullable=False
    )
    date_of_birth: Mapped[date | None] = mapped_column(Date, nullable=True)
    weight_kg: Mapped[float | None] = mapped_column(Float, nullable=True)
    color: Mapped[str | None] = mapped_column(String(80), nullable=True)
    microchip_id: Mapped[str | None] = mapped_column(String(80), nullable=True)
    is_neutered: Mapped[bool] = mapped_column(default=False, nullable=False)
    notes: Mapped[str | None] = mapped_column(String(2000), nullable=True)
    avatar_url: Mapped[str | None] = mapped_column(String(500), nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=_utcnow, nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=_utcnow, onupdate=_utcnow, nullable=False
    )

    owner = relationship("User", backref="pets")

File: client/types/pet.ts
export type PetSpecies =
  | "dog"
  | "cat"
  | "bird"
  | "rabbit"
  | "fish"
  | "reptile"
  | "rodent"
  | "other";

export type PetGender = "male" | "female" | "unknown";

export interface Pet {
  id: string;
  ownerId: string;
  name: string;
  species: PetSpecies;
  breed: string | null;
  gender: PetGender;
  dateOfBirth: string | null;
  weightKg: number | null;
  color: string | null;
  microchipId: string | null;
  isNeutered: boolean;
  notes: string | null;
  avatarUrl: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface PetListItem {
  id: string;
  name: string;
  species: PetSpecies;
  breed: string | null;
  dateOfBirth: string | null;
  avatarUrl: string | null;
}

interface RawPet {
  id: string;
  owner_id: string;
  name: string;
  species: PetSpecies;
  breed: string | null;
  gender: PetGender;
  date_of_birth: string | null;
  weight_kg: number | null;
  color: string | null;
  microchip_id: string | null;
  is_neutered: boolean;
  notes: string | null;
  avatar_url: string | null;
  created_at: string;
  updated_at: string;
}

interface RawPetListItem {
  id: string;
  name: string;
  species: PetSpecies;
  breed: string | null;
  date_of_birth: string | null;
  avatar_url: string | null;
}

export function mapPet(raw: RawPet): Pet {
  return {
    id: raw.id,
    ownerId: raw.owner_id,
    name: raw.name,
    species: raw.species,
    breed: raw.breed,
    gender: raw.gender,
    dateOfBirth: raw.date_of_birth,
    weightKg: raw.weight_kg,
    color: raw.color,
    microchipId: raw.microchip_id,
    isNeutered: raw.is_neutered,
    notes: raw.notes,
    avatarUrl: raw.avatar_url,
    createdAt: raw.created_at,
    updatedAt: raw.updated_at,
  };
}

export function mapPetListItem(raw: RawPetListItem): PetListItem {
  return {
    id: raw.id,
    name: raw.name,
    species: raw.species,
    breed: raw.breed,
    dateOfBirth: raw.date_of_birth,
    avatarUrl: raw.avatar_url,
  };
}
